### PR TITLE
[TS migration] Migrate 'SettingsAppDownloadLinks' page to TypeScript #25175

### PR DESCRIPTION
**Pull request by hkopser99**
### Details 

-Migrated JSX component to Typescript 

-Extended MenuItemProps type as DownloadMenuItem type  

-Typed the MenuItems as an array of DownloadMenuItem

-Removed withWindowDimensions/withLocalize PropTypes and compose() from export statement per guidelines on migration  

-Used Hooks for localize translations

-Removed underscore in JSX mapping of menu items. I saw other files converted to TSX that removed it and were getting ESLint warnings. Using the _ is noted as a best practice but given other TSX files had removed or used an ESLint ignore I followed suit. I did not get ESLint warnings, so I did not need the ignore comment. 

-I removed onKeyDown attribute and blur() from MenuItem. I did not see this elsewhere in the codebase and the removal had no impact on the page’s functionality or appearance. It appeared antiquated but please advise if there is a need for it that I am missing)  

-This component receives props from Navigation but I do not see elsewhere in the codebase that we are explicitly listing navigation related props when migrating to TS (nav/routing props should inherently apply to all pages presented via Navigation) 

-A TS warning made me list translationKey in DownloadMenuItem as type TranslationPaths. The warning only appeared when not providing translate from useLocalize() with a string literal. Others working on TS migrations may want to be aware of this as a solution. 

-No other components or other files in the codebase import this file so changing from .js to .tsx should not require revisions to import statements. This component is routed to by using its displayName and that was not changed. 

-One item to note is that on the iOS simulator there is no App Store app, so it does not know how to handle the download iOS app link click (see screenshots on iOS Native and iOS mWeb for reference). The link supplied is correct, and it works everywhere else.

 

### Fixed Issues 
$ #25175 

PROPOSAL:
$ https://github.com/Expensify/App/issues/25175#issuecomment-1884662977  

Proposed to migrate file to Typescript and convert file type to TSX following Expensify migration best practices as well as testing on all platforms to confirm page functionality remains the same. 

 
### Tests 

I ran the app on each platform, confirmed each App Download Link leads to the intended URL. I also confirmed that navigation back to the About page operates as intended when leaving the AppDownloadLinks page. 
  

- [X] Verify that no errors appear in the JS console 

  

### Offline tests 

The links work offline but will not present web pages with content or the MacOS App Download option until internet connection has been reestablished. During this time standard Safari/Chrome warnings are presented. 
  

### QA Steps 

1. Login and then visit Settings->About->App Download Links
2. Click/tap each link to confirm they are sending the user to the right download website (e.g. Google Play, App Store/Apple Store website, or desktop download for MacOS App)
3. Navigate back to the About page when done to confirm navigation functions as intended
  

- [X] Verify that no errors appear in the JS console 

  

### PR Author Checklist 

 

  

- [X] I linked the correct issue in the `### Fixed Issues` section above 

- [X] I wrote clear testing steps that cover the changes made in this PR 

    - [X] I added steps for local testing in the `Tests` section 

    - [X] I added steps for the expected offline behavior in the `Offline steps` section 

    - [X] I added steps for Staging and/or Production testing in the `QA steps` section 

    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct) 

    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline) 

    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability). 

- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms) 

- [X] I ran the tests on **all platforms** & verified they passed on: 

    - [X] Android: Native 

    - [X] Android: mWeb Chrome 

    - [X] iOS: Native 

    - [X] iOS: mWeb Safari 

    - [X] MacOS: Chrome / Safari 

    - [X] MacOS: Desktop 

- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed) 

- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code)) 

    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`) 

    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`. 

    - [X] I verified that comments were added to code that is not self explanatory 

    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing. 

    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60) 

      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: 

    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68) 

    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy. 

    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README. 

    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed 

- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers 

- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md) 

- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected) 

- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests) 

- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such 

- [X] I verified that if a function's arguments changed that all usages have also been updated correctly 

- [X] If any new file was added I verified that: 

    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory 

- [X] If a new CSS style is added I verified that: 

    - [X] A similar style doesn't already exist 

    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`) 

- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic. 

- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases) 

- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected. 

- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account. 

- [X] If the PR modifies the form input styles: 

    - [X] I verified that all the inputs inside a form are aligned with each other. 

    - [X] I added `Design` label so the design team can review the changes. 

- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page. 

- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps. 

  

### Screenshots/Videos 

<details> 

<summary>Android: Native</summary> 

  

ExpensifyADLTSMigAndroidEmulatorPage 

ExpensifyADLTSMigAndroidEmulatorAndroidAppLink 

ExpensifyADLTSMigAndroidEmulatoriOSAppLink 

ExpensifyADLTSMigAndroidEmulatorMacOSAppLink 

</details> 

  

<details> 

<summary>Android: mWeb Chrome</summary> 

  

ExpensifyADLTSMigChromeWebPage 

ExpensifyADLTSMigChromeWebGooglePlayLink 

ExpensifyADLTSMigChromeWebiOSApp 

ExpensifyADLTSMigChromeWebMacOSLink 

  

</details> 

  

<details> 

<summary>iOS: Native</summary> 

  

ExpensifyADLTSMigiOSNative 

ExpensifyADLTSMigiOSNativeAndroidLink 

ExpensifyADLTSMigiOSNativeiOSAppLink 

ExpensifyADLTSMigiOSNativeMacOSLink 

  

</details> 

  

<details> 

<summary>iOS: mWeb Safari</summary> 

  

ExpensifyADLTSMigmSafariPage 

ExpensifyADLTSMigMSafariAndroidLink 

ExpensifyADLTSMigMSafariiOSAppLink 

ExpensifyADLTSMigMSafariMacOSAppLink 

  

</details> 

  

<details> 

<summary>MacOS: Chrome / Safari</summary> 

  

ExpensifyADLTSMigDesktopPg 

ExpensifyADLTSMigDesktopLinksWorking 

  

</details> 

  

<details> 

<summary>MacOS: Desktop</summary> 

  

ExpensifyADLTSMigMacOSAppPage 

ExpensifyADLTSMigMacOSAppLinksWorking 

  

  

</details>

### CODE REVIEW

**Comment by hkopser99** [comment]
I have read the CLA document and I hereby sign the CLA

**Comment by hkopser99** [comment]
recheck

**Comment by Santhosh-Sellavel** [comment]
Can you fix the lint issues?

**Comment by Santhosh-Sellavel** [comment]
Please link the issue properly in the description. Also, can you check and update descriptions correctly there few typos

**Comment by hkopser99** [comment]
@Santhosh-Sellavel ran Prettier so the lint issues should be resolved. Do I need to tell the bot to recheck as I did after signing the agreement? I also fixed the description and the linking to the fix along with a few other revisions for clarity and conciseness. I found another recent PR and tried to make mine more like that so you have an easier review. If you notice anything else missing just let me know. Its my first PR for Expensify so I want to make sure you have everything you need.

**Comment by Santhosh-Sellavel** [comment]
@Julesssss 
Can you initiate workflow checks?

**Comment by hkopser99** [comment]
@Santhosh-Sellavel I saw an error occurred so I ran the perf-tests locally and didn't see it on my local environment. I found a few different threads in slack from Nov 2023 - [Feb](https://expensify.slack.com/archives/C01GTK53T8Q/p1706732738472299) of this year mentioning perf-test failures where changes were unrelated to the failure. At least one was resolved just by re-running the tests. Is this an issue you've encountered before?

**Comment by kubabutkiewicz** [comment]
@hkopser99 for fixing Reassure Perfomance tests just marge main to your branch

File: src/pages/settings/AppDownloadLinks.tsx [hunk 1]
```diff
@@ -1,75 +1,75 @@
...
-};
-
-function AppDownloadLinksPage(props) {
+function AppDownloadLinksPage() {
     const styles = useThemeStyles();
-    let popoverAnchor;
+    const {translate} = useLocalize();
+    const popoverAnchor = useRef<View>(null);
+
+    type DownloadMenuItem = MenuItemProps & {
```
[the remark this answers is not in the document]
Comment (hkopser99) [reply src/pages/settings/AppDownloadLinks.tsx:23]: Thanks, I'll move that here and will be sure to move type declarations outside of component scope going forward.

File: src/pages/settings/AppDownloadLinks.tsx [hunk 2]
```diff
@@ -1,75 +1,75 @@
...
         },
     ];
 
     return (
         <ScreenWrapper testID={AppDownloadLinksPage.displayName}>
             <HeaderWithBackButton
-                title={props.translate('initialSettingsPage.aboutPage.appDownloadLinks')}
-                onBackButtonPress={() => Navigation.goBack()}
+                title={translate('initialSettingsPage.aboutPage.appDownloadLinks')}
+                onBackButtonPress={() => Navigation.goBack(ROUTES.SETTINGS_ABOUT)}
```
[the remark this answers is not in the document]
Comment (hkopser99) [reply src/pages/settings/AppDownloadLinks.tsx:63]: My logic when I was working on the change was to make the navigation destination explicit. Now that I look at it there's no benefit to this. Navigation.goBack() works as intended here and removing the explicit destination makes the code more reusable if we ever want to present the AppDownloadLinks page from another screen in the future. If approved I will amend to remove ROUTES.SETTINGS_ABOUT as an argument and will then also be able to remove the ROUTES import statement above as this was the only use of ROUTES.

**Comment by Santhosh-Sellavel** [comment]
@Julesssss Can you rerun workflow tests

**Comment by Julesssss** [comment]
> @Julesssss Can you rerun workflow tests

Yep, done

File: src/pages/settings/AppDownloadLinks.tsx [hunk 3]
```diff
@@ -1,75 +1,75 @@
...
         {
             translationKey: 'initialSettingsPage.appDownloadLinks.android.label',
-            icon: Expensicons.Android,
-            iconRight: Expensicons.NewWindow,
-            action: () => {
+            openAppDownloadLink: () => {
                 Link.openExternalLink(CONST.APP_DOWNLOAD_LINKS.ANDROID);
             },
-            link: CONST.APP_DOWNLOAD_LINKS.ANDROID,
+            downloadLink: CONST.APP_DOWNLOAD_LINKS.ANDROID,
```
Comment (situchan) [inline src/pages/settings/AppDownloadLinks.tsx:35]: Any reason for renaming? action > openAppDownloadLink, link > downloadLink

File: src/pages/settings/AppDownloadLinks.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (situchan) [reply src/pages/settings/AppDownloadLinks.tsx:63]: Keep original logic if possible.

**Review by situchan — CHANGES_REQUESTED** [review]
Please merge main as 1k+ commits are behind

**Comment by situchan** [comment]
@hkopser99 I am not sure what happened with your author checklist. Please check any other PRs

Screenshot 2024-02-14 at 3 32 26 AM


---

Add Tests / QA Steps

File: src/pages/settings/AppDownloadLinks.tsx (hunk 3 above)
Comment (hkopser99) [reply src/pages/settings/AppDownloadLinks.tsx:35]: Yes, I was following my understanding of the Style document in the Naming Conventions->Event Handlers section found [here](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#naming-conventions). I then adjusted link to downloadLink to be more descriptive. Please let me know if 'action' and 'link' are preferred to the more descriptive names.

Comment (situchan) [reply src/pages/settings/AppDownloadLinks.tsx:35]: original names are fine to me

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by Julesssss** [comment]
Hey @situchan, no rush but I think we're just waiting on the final device testing and screen captures, right?

**Comment by situchan** [comment]
> Hey @situchan, no rush but I think we're just waiting on the final device testing and screen captures, right?

right

**Review by situchan — APPROVED** [review]
:shipit:

**Review by Julesssss — APPROVED** [review]

**Comment by Julesssss** [comment]
Hi @hkopser99 would you mind merging `main` again to confirm there aren't any conflicts please

**Comment by hkopser99** [comment]
@Julesssss just merged!

**Comment by Julesssss** [comment]
Thanks

**Comment by hkopser99** [comment]
@Julesssss thank you and thanks for letting me work on this one. Had some bumps along the way getting up and running w/ the codebase/dependencies and my laptop but thankfully the actual migration piece wasn't tricky :) Do you have any more of these or other stuff to work on? Would love to do more

**Comment by Julesssss** [comment]
No worries! You can find all our open issues [here](https://github.com/Expensify/App/issues?q=is%3Aissue+label%3AExternal+is%3Aopen), but some of them will have already been assigned to contributors.

**Pull request merged**